src/k2/cpo/client/Client.h

```diff
@@ -202,54 +220,76 @@ class CPOClient {
         return RPC().callRPC<dto::MetadataGetRequest, dto::MetadataGetResponse>(dto::Verbs::CPO_PARTITION_METADATA_GET, request, *cpo, timeout);
     }
 
-    ConfigDuration partition_request_timeout{"partition_request_timeout", 100ms};
-    ConfigDuration schema_request_timeout{"schema_request_timeout", 1s};
-    ConfigDuration cpo_request_timeout{"cpo_request_timeout", 100ms};
-    ConfigDuration cpo_request_backoff{"cpo_request_backoff", 500ms};
 private:
-    template <typename ClockT=Clock>
-    seastar::future<std::tuple<Status, seastar::lw_shared_ptr<dto::PartitionGetter>>>
-    _getPartitionGetterForCollection(Deadline<ClockT> deadline, const String& cname, const dto::Key& key=dto::Key{},
+    std::unordered_map<String, std::vector<seastar::promise<Status>>> _requestWaiters;
+    void _fulfillWaiters(const String& name, const Status& status);
+
+    // Check if all partitions are assigned within the deadline. Returns a future of assignment status.
+    template<typename ClockT=Clock>
+    seastar::future<Status> _checkAllParititonsAssigned(Deadline<ClockT> deadline, const String& name) {
+        return _getAssignedPartitionWithRetry(deadline, name, dto::Key{})
+        .then([this, name, deadline](Status&& status){
+            K2LOG_D(log::cpoclient, "Checking if all partitions are assigned");
+            if (status.is2xxOK()) {
+                // Then check if all partitions are assigned
+                for (auto it = collections.find(name); it != collections.end(); ++it) {
+                    for (auto& partition : it->second->getAllPartitions()) {
+                        if (partition.astate == dto::AssignmentState::NotAssigned) {
```

Comment:
Should check if it is Assigned here since it could be in other states like pending.